src/query/server/server.go

```diff
@@ -114,17 +118,111 @@ func Run(runOpts RunOptions) {
 	logger := logging.WithContext(ctx)
 	defer logger.Sync()
 
-	scope, _, err := cfg.Metrics.NewRootScope()
+	scope, closer, err := cfg.Metrics.NewRootScope()
 	if err != nil {
 		logger.Fatal("could not connect to metrics", zap.Any("error", err))
 	}
 
+	// Close metrics scope
+	defer func() {
+		logger.Info("closing metrics scope")
+		if err := closer.Close(); err != nil {
+			logger.Error("unable to close metrics scope", zap.Error(err))
+		}
+	}()
+
+	var (
+		backendStorage storage.Storage
+		clusterClient  clusterclient.Client
+		downsampler    downsample.Downsampler
+		enabled        bool
+	)
+
+	// For grpc backend, we need to setup only the grpc client and a storage accompanying that client.
+	// For m3db backend, we need to make connections to the m3db cluster which generates a session and use the storage with the session.
+	if cfg.Backend == config.GRPCStorageType {
+		backendStorage, enabled, err = remoteClient(cfg)
+		if err != nil {
+			logger.Fatal("unable to setup grpc backend", zap.Error(err))
+		}
+		if !enabled {
+			logger.Fatal("need remote clients for grpc backend")
+		}
+
+		logger.Info("setup grpc backend")
+	} else {
+		var cleanup cleanupFn
+		backendStorage, clusterClient, downsampler, cleanup, err = newM3DBStorage(runOpts, cfg, logger, scope)
+		if err != nil {
+			logger.Fatal("unable to setup m3db backend", zap.Error(err))
+		}
+		defer cleanup()
+	}
+
+	engine := executor.NewEngine(backendStorage)
+
+	handler, err := httpd.NewHandler(backendStorage, downsampler, engine,
+		clusterClient, cfg, runOpts.DBConfig, scope)
+	if err != nil {
+		logger.Fatal("unable to set up handlers", zap.Error(err))
+	}
+
+	if err := handler.RegisterRoutes(); err != nil {
+		logger.Fatal("unable to register routes", zap.Error(err))
+	}
+
+	listenAddress, err := cfg.ListenAddress.Resolve()
+	if err != nil {
+		logger.Fatal("unable to get listen address", zap.Error(err))
+	}
+
+	srv := &http.Server{Addr: listenAddress, Handler: handler.Router}
+	defer func() {
+		logger.Info("closing server")
+		if err := srv.Shutdown(ctx); err != nil {
+			logger.Error("error closing server", zap.Error(err))
+		}
+
+	}()
+
+	go func() {
+		logger.Info("starting server", zap.String("address", listenAddress))
+		if err := srv.ListenAndServe(); err != nil {
+			logger.Error("server error while listening",
+				zap.String("address", listenAddress), zap.Error(err))
+		}
+	}()
+
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, os.Interrupt, syscall.SIGTERM)
+
+	var interruptCh <-chan error = make(chan error)
+	if runOpts.InterruptCh != nil {
+		interruptCh = runOpts.InterruptCh
+	}
+
+	select {
+	case <-sigChan:
+	case <-interruptCh:
+	}
+}
+
+// make connections to the m3db cluster(s) and generate sessions for those clusters along with the storage
+func newM3DBStorage(
+	runOpts RunOptions,
+	cfg config.Configuration,
+	logger *zap.Logger,
+	scope tally.Scope,
+) (storage.Storage, clusterclient.Client, downsample.Downsampler, cleanupFn, error) {
```

Comment:
nit: 5 return args makes it a bit harder to keep track of what's going on. Your call to change or leave it

Reply:
yeah unfortunately it was initially structured as a monolith. I didn't want to spend time fixing it :(